src/libraries/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs

```diff
@@ -465,24 +465,34 @@ internal static string JavaScriptStringEncode(string? value)
             return expandedBytes;
         }
 
-        //  Helper to encode the non-ASCII url characters only
+        //  Helper to encode the non-ASCII url characters only or spaces
         private static string UrlEncodeNonAscii(string str, Encoding e)
         {
             Debug.Assert(!string.IsNullOrEmpty(str));
             Debug.Assert(e != null);
+
+            const int StackallocThreshold = 512;
+
+            if (e.GetMaxByteCount(str.Length) <= StackallocThreshold)
+            {
+                Span<byte> byteSpan = stackalloc byte[StackallocThreshold];
+                int encodedBytes = e.GetBytes(str, byteSpan);
+
+                return UrlEncodeNonAscii(byteSpan.Slice(0, encodedBytes));
+            }
             byte[] bytes = e.GetBytes(str);
-            byte[] encodedBytes = UrlEncodeNonAscii(bytes, 0, bytes.Length);
-            return Encoding.ASCII.GetString(encodedBytes);
+            return UrlEncodeNonAscii(bytes.AsSpan(0, bytes.Length));
         }
 
-        private static byte[] UrlEncodeNonAscii(byte[] bytes, int offset, int count)
+        private static string UrlEncodeNonAscii(Span<byte> bytes)
```

Comment:
These methods are still allocating quite a bit (the original string even if it's already Ascii, substring and combined string in `UrlPathEncodeImpl` ...).

The 1 caller of this helper is using `Encoding.UTF8` and wants to stop after you see a `?`.
Assuming there's sufficient test coverage, I would change this to instead look something like

```c#
private static string UrlPathEncodeImpl(string value)
{
    int i = value.AsSpan().IndexOfAnyExceptInRange((char)0x21, (char)0x7F);
    if (i < 0)
    {
        return value;
    }

    int indexOfQuery = value.IndexOf('?');
    if (indexOfQuery >= 0 && indexOfQuery < i)
    {
        // Everything before the Query is valid ASCII
        return value;
    }

    ReadOnlySpan<char> toEncode = indexOfQuery >= 0
        ? value.AsSpan(i, indexOfQuery - i)
        : value.AsSpan(i);

    byte[] bytes = ArrayPool<char>.Shared.Rent(Encoding.UTF8.GetMaxByteCount(toEncode.Length));
    char[] chars = ArrayPool<char>.Shared.Rent(bytes.Length * 3);

    int utf8Length = Encoding.UTF8.GetBytes(toEncode, bytes);
    int charCount = 0;
    foreach (byte b in bytes.AsSpan(0, utf8Length))
    {
        if (IsNonAsciiOrSpaceByte(b))
        {
            chars[charCount++] = '%';
            chars[charCount++] = HexConverter.ToCharLower(b >> 4);
            chars[charCount++] = HexConverter.ToCharLower(b);
        }
        else
        {
            chars[charCount++] = (char)b;
        }
    }

    ArrayPool<byte>.Shared.Return(bytes);
    string result = string.Concat(
        value.AsSpan(0, i),
        chars.AsSpan(0, charCount),
        indexOfQuery >= 0 ? value.AsSpan(indexOfQuery) : ReadOnlySpan<char>.Empty);
    ArrayPool<char>.Shared.Return(chars);
    return result;
}
```

This would avoid a bunch of intermediate allocations and transcoding.